### DISCUSSION TITLE
Editing/erasing within imported handwritten SVG/PDF notes

### DISCUSSION THREAD

**Discussion by tobiaskramer** [flxzt/rnote#733] [Ideas]
I just started using rnote. It looks fantastic and has the potential to entirely replace my OneNote based workflow. 

However, I have many existing documents (handwritten lecture notes, sometimes with images/urls/small text boxes) and need to erase for instance a small mistake in an equation or shift a (handwritten) word/letter around. Once I am able to do this within rnote, I can work completely within rnote for the future.

So far I did not manage to do the importing and editing of handwritten letters/equations using either PDF or SVG  import to rnote.
Here are my two attempted workflows:

1. OneNote export PDF -> Import into rnote. I get the imported OneNote Page as a vector image in rnote, not able to erase some of my handwritten strokes

2. OneNote export PDF -> Import into Inkscape -> Export from Inkscape as SVG (I tried plain SVG, Inkscape SVG, with/without ungroup all elements extension). In Inkscape I can move/erase etc individual elements of my handwriting -> Import SVG into rnote

I get then again a "Vector Image" (this time with a transparent background as compared to (1.), but again I cannot erase/move/edit individual elements.

Am I missing something obvious, or is this a planned future enhancement?

I would ideally imagine to check a box upon import to "make all strokes/elements editable within rnote" rather than obtaining a single "vector image".

Thanks for this fantastic software, I think it has the potential to allow us to move beyond OneNote and so far looks nicer to me than many other approaches!

-Tobias

**Comment by flxzt** [comment]
Thanks for the kind words!

Importing handwriting from other note-taking apps is complicated. As you already discovered, Pdf pages and Svg's are currently imported as non-editable vector images. The only format where a "richer" import is supported is `.xopp` files from Xournal++, because the structure of the data is similar to the one used inside rnote (strokes which have have a clear defined path with coordinates, pressure values, a color).

The problems we are faced with when trying to deconstruct Pdf's and Svg's are hard to solve (I'd say out of scope for a "simple" note taking app like rnote):
- we'd need to essentially support all of the Svg spec (gradients, paths, etc.) to be able to display and edit them (like Inkscape)
- when note taking apps are exporting to Pdf or Svg, the stroke data gets lost, because Svg unfortunately does not support variable width strokes. I think OneNote exports regular strokes as `<path>`s filled with a solid color made up of curved elements. I haven't tried it, but it is likely very difficult to recognize if a shape is supposed to be handwriting and then extract path and pressure information from it, even on a best fit/effort basis.

So, that leaves us with the following possible enhancements (that I can think of):

- deconstruct Svg's only partially, like into individual images for every  top-level svg group or something like that. That could be a good way to make content "more" editable in a general way. The downside of that it is difficult to predict how fine-grained this deconstruction should be to make it work for most content.
- add custom imports for more handwriting app formats: at least OneNote does have a [specification](https://interoperability.blob.core.windows.net/files/MS-ONE/%5bMS-ONE%5d.pdf), so that should be possible. For other formats it is more difficult and adds to maintainer burden.

I currently can't offer you a way to solve your problem, but I hope at least that this clears things up a bit!

**Reply by tobiaskramer** [reply]
Thanks for the detailed reply and explanation. I figured out the only working way out of OneNote:
there is a firefox extension

https://addons.mozilla.org/en-US/firefox/addon/onenote-to-xournal/

which intercepts the OneNote drawing calls in the web browser version of OneNote and saves the strokes (handwriting) and images as Xjournal++ .xopp file.

The OneNote -> Xournal++ conversion does not work if there  are certain elements present (i.e. links) and does not export the text fonts for my notes correctly. But it works well enough for me after cleaning my OneNote pages (i.e. removing links) to be able to export them with the firefox extension and then import and edit them in Xournal++. 
(There exists also a standalone program for OneNote -> .xopp conversion which did not work for any of my files though).

Then I opened the .xopp file in Xournal++ and they are fine (I can edit etc). 

When I open the exported OneNote -> .xopp file in rnote some elements are not quite right, I can file separate bug reports if warranted:

- some boxes (rectangles) are shown with only three edges instead of 4
- somehow the page size does not  show correctly in rnote (it is set to A4/A5 in Xournal++). 
In rnote it becomes custom format.
- In rnote I cannot afterwards change the page format to A4/A5 and I cannot access the Landscape/Portrait orientation buttons (none of the two choices is selectable) after importing the .xopp file.

These are relatively large files (about 30-40 mb .xopp file, ca 40 pages handwritten notes with some images).

Thanks again for your time to develop this nice program! I will see if I can use it and would appreciate in particular if it maintains some interoperability with Xournal++ {for instance some limited export rnote -> Xournal++)

**Reply by flxzt** [reply]
The extension is interesting, but from brief testing and looking at it's source code it seems that it is only able to convert strokes with constant width. OneNote Web apparently is also unable to import documents created with the desktop app, which is odd.

The way the extension does the conversion is by inspecting the web content, not the file itself. I doubt it will ever support converting variable width strokes. Here is a screenshot that shows the generated content of OneNote web:

![Screenshot 2023-07-11 141421](https://github.com/flxzt/rnote/assets/19841886/070ed7ae-1b5a-42fa-bc40-d9069adc52f3)

Regarding the issues:

- the missing side of closed shapes was just a rendering bug, I just pushed a fix (908f6ed). The shapes themselves get imported correctly.
- Regarding the page format presets: Yeah, that is a current limitation of rnote, because only the actual size is stored, not the selected preset (because the size values vary with different Dpi when the preset format is applied.).
- Making the portrait/landscape toggle work with "custom format" would be a good improvement, I agree. We could simply check if `width > height` and update it's state based on that.

> Thanks again for your time to develop this nice program! I will see if I can use it and would appreciate in particular if it maintains some interoperability with Xournal++ {for instance some limited export rnote -> Xournal++)

Definitely! Good interoperability and providing import/export for a variety of formats is a focus and important to me personally.
